primitives/src/transaction.rs

```diff
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Bitcoin transactions.
+//!
+//! A transaction describes a transfer of money. It consumes previously-unspent
+//! transaction outputs and produces new ones, satisfying the condition to spend
+//! the old outputs (typically a digital signature with a specific key must be
+//! provided) and defining the condition to spend the new ones. The use of digital
+//! signatures ensures that coins cannot be spent by unauthorized parties.
+//!
+//! This module provides the structures and functions needed to support transactions.
+
+use hashes::sha256d;
+
+hashes::hash_newtype! {
+    /// A bitcoin transaction hash/transaction ID.
+    ///
+    /// For compatibility with the existing Bitcoin infrastructure and historical and current
+    /// versions of the Bitcoin Core software itself, this and other [`sha256d::Hash`] types, are
+    /// serialized in reverse byte order when converted to a hex string via [`std::fmt::Display`]
+    /// trait operations.
+    ///
+    /// See [`hashes::Hash::DISPLAY_BACKWARD`] for more details.
+    pub struct Txid(sha256d::Hash);
+
+    /// A bitcoin witness transaction ID.
+    pub struct Wtxid(sha256d::Hash);
```

Comment:
Note that we still have one significant problem here: These implement `FromStr` which publicly exposes unstable `Err` type and `Hash` trait from an unstable crate. I'm not going to block the PR over this but we absolutely have to resolve this before `primitives` stabilizes.

Reply:
I'm not sure but I was mainly thinking of wrapping the errors in `struct ParseTxidError(hex::Error);` at least temporarily.

Reply:
Doing it temporarily is just as bad (for API stability) as exposing a type from an unstable crate. And wrapping the error is pretty bad for consumers -- if they are parsing a bunch of data from hex they will then need to crate a new enum to hold all the type-specific wrapped errors, thus creating a massive type that ultimately only holds hex errors!

Another thing we could do is stabilize just the hex error types -- and promise to semver-trick them forever until 1.0.

Reply:
> Doing it temporarily is just as bad (for API stability)

How? Exposing an error is not breaking change outside of stupid case when the error is an enum (I plan to not make it enum) and someone writes `let ParseTxidError { .. } = "foo".parse::<Txid>().unwrap_err();`

> crate a new enum to hold all the type-specific wrapped errors, thus creating a massive type that ultimately only holds hex errors!

That's a good point. While it's not too bad from performance perspective and might have advantages, forcing users to do it is excessive. We could just expose one wrapped error type for now but I think it'd have to be per-crate error type which is annoying.

> Another thing we could do is stabilize just the hex error types -- and promise to semver-trick them forever until 1.0.

That doesn't really work that well. Aside from semver trick having certain downsides, I think we would need to also stabilize at least some decoding functions (at least one) that would produce the error. Otherwise we would have to make the error fully public and final. Maybe we should have `minimal-hex-decode` crate that's 1.0 and only contains decoding and error types.

Reply:
Looking at `hex-conservative`, we'll have to remove the `core2` feature to stabilize and fix the char handling in error types at least.

I've also realized, in the first iteration we could also not have `FromStr` but that'd get many people pissed and I wouldn't be happy about it either.

Reply:
> How? Exposing an error is not breaking change outside of stupid case when the error is an enum (I plan to not make it enum) and someone writes `let ParseTxidError { .. } = "foo".parse::<Txid>().unwrap_err();`

When we replace the returned type with the actual type, that will be a breaking change.

>Maybe we should have minimal-hex-decode crate that's 1.0 and only contains decoding and error types.

Lol maybe we should call it `hex-maximally-conservative`.

Reply:
> When we replace the returned type with the actual type, that will be a breaking change.

Technically yes, but the breakage is so crazy obscure and so easy to fix/avoid that it shouldn't really be an issue. Maybe there's also a hack to straight up abort compilation if you're getting into a situation in which you theoretically could get to such situation.

The only case when it's breaking is if the consumer starts using `hex-conservative` 1.0, potentially updating the minor version and does **not** update `primitives` to use the new `hex-conservative` and then implements the same trait for both `primitives::ParseHashError` and `hex-conservative::HexToArrayError` and then later upgrades `primitives`. The solution is obviously to delete one of the impls.

A rule to avoid this is: once `hex-conservative` 1.0 is out you must update all the crates that depended on `hex-conservatives` and exposed hex errors. Any hack that would attempt to abort compilation would need to abort if the presence of `hex-conservative` 1.0 is detected in the tree somehow. Though I'm afraid doing so cleanly would require another dummy helper crate.

Reply:
Ok, makes sense. I'd be willing to accept such an obscure form of breakage.

TBH I'd also be willing to accept a permanent wrapper, though once `hex-conservative` is stable I'd like to make the inner field public.

As for "one wrapper or multiple" I'd rather we just have one wrapper which we call `ParseHashError` or something.

Reply:
I figured out a hack to somewhat cleanly enforce the upgrade (but I should really test what cargo does in that situation):

We release a `hex-conservative-version-check` crate that's literally empty aside from doc explaining its purpose. Then we depend on it from `primitives` with `version = "=1.0.0". Once `hex-conservative` 1.0 is ready we'll release `hex-conservative-version-check` 1.1 and depend on it from `hex-conservative` 1.0 and we'll drop/update the dependency in `primitives` 1.x which introduces the dependency on `hex-conservative` 1.0

Though I wouldn't be surprised if it's considered too much hassle for little benefit.

Reply:
I had it in mind that all our leaf crates excluding `internals` would stabalise before `primitives` did, that is kind of why we have a perma-unstable `internals` crate, right?

Said another way:

Can we just grow a pair of balls and stabalize at least one crate?

Reply:
>Though I wouldn't be surprised if it's considered too much hassle for little benefit.

Yes, far too much hassle. If a sequence of `cargo update -p` commands fixes the breakage we'll just tell users to do that. We already have to do that to deal with the constant MSRV breakage from our dependencies.

Reply:
> I had it in mind that all our leaf crates excluding `internals` would stabalise before `primitives` did

I didn't because it'd take obscene amount of time for very little benefit. The main focus now is to get different crates across the ecosystem to work together without constant **coordinated** upgrades. There should be ideally no types exposed from `hex-conservative` or `hashes` in the public API and the only thing that throws a wrench into this plan are the parsing errors.

> Can we just grow a pair of balls and stabalize at least one crate?

Doing something you know is 100% wrong isn't "growing pair of balls". That's just insanity.

Reply:
What's 100% wrong is if we have planned changes and we stabilize anyway.

>  But if it were between stabilizing hex-conservative and failing to stabilize anything, we should obviously stabilize hex-conservative.

Yes, but that's a false dichotomy that we're not experiencing. We have at least two other options.